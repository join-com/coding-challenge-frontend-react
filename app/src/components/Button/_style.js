import styled from 'styled-components';

export const Button = styled.button`
    font-weight: 400;
    text-transform: uppercase;
    margin: 0;
    cursor: pointer;
    border: 0;
    background: none;
    padding: 0 10px;
    color: #fff;
    background-color: #1187d2;
`;

export const Link = Button.withComponent('a');
