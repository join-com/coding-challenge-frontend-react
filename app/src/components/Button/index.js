import React from 'react';
import PropTypes from 'prop-types';

import {
    Link,
    Button as ButtonStyledComponent
} from './_style';

export default function Button({
    tag,
    children,
    ...attrs
}) {
    const Component = tag === 'a' ? Link : ButtonStyledComponent;

    return <Component {...attrs}>{ children }</Component>;
}

Button.propTypes = {
    type: PropTypes.oneOf(['reset', 'button', 'submit']),
    tag: PropTypes.oneOf(['a', 'button']),
    className: PropTypes.string,
    onClick: PropTypes.func,
    children: PropTypes.node,
    disabled: PropTypes.bool
};

Button.defaultProps = {
    tag: 'button',
    type: 'button'
};
