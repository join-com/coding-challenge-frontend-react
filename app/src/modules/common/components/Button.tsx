import * as React from 'react'

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement>

export const Button: React.FC<ButtonProps> = ({ children, ...props }) => {
  return (
    <button {...props}>
      <style jsx={true}>
        {/* language=SCSS */ `
          button {
            background-color: #0755ff;
            border: none;
            border-radius: 0;
            outline: none;
            color: #fff;
            cursor: pointer;
            padding: 8px 15px;
            height: 36px;
            font-size: 14px;

            &:hover:not(:focus) {
              background-color: #0640bf;
            }

            &:active {
              box-shadow: 0 3px 6px -2px inset #333;
            }

            &:focus:not(:active) {
              padding: 6px 13px;
              border: 2px solid #053499;
            }
          }
        `}
      </style>
      {children}
    </button>
  )
}
