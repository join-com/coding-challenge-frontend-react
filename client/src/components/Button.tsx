import React, { FC } from "react";
import Styled from "styled-components";

interface IButton {
  className?: string;
  children?: React.ElementType[];
}

const Button: FC<IButton> = ({ className, children }) => <button>{children}</button>;

export default Styled(Button)``;
