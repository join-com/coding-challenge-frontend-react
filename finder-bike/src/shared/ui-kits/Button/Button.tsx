import React from 'react';
import styled from 'styled-components';
import * as Colors from '../Variables/Colors';

export type ButtonColor = 'success' | 'warning' | 'danger' | 'primary';

export interface IButtonProps {
  children?: React.ReactNode;
  onClick?: (e: any) => void;
  color: ButtonColor;
  disabled?: boolean;
  type?: string;
  isFull?: boolean;
}

const StyledButton = styled.button<IButtonProps>`
  color: ${Colors.white};
  background-color: ${(props: IButtonProps) => Colors[props.color]};
  border-color: ${(props: IButtonProps) => Colors[props.color]};
  outline: none;

  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1.3rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  ${(props: IButtonProps) =>
    props.disabled && `background-color: ${Colors.grey};`}
  cursor: pointer;

  ${({ isFull }) => isFull && 'width: 100%'}
`;

export default function Button(props: IButtonProps) {
  return <StyledButton {...props} />;
}
