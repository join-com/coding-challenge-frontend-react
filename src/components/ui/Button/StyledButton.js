import styled, { css } from "styled-components";
import { colors } from "../../../constants/styles";
import { defaultInputStyles } from "../../../helpers/stylesMixins";

const buttonHoverStyles = css`
  color: ${colors.primary};
`;

const StyledButton = styled.button`
  ${defaultInputStyles};
  text-align: center;
  margin-right: 0.5rem;
  &:hover {
    ${({ disabled }) => !disabled && buttonHoverStyles};
  }
`;

export default StyledButton;
