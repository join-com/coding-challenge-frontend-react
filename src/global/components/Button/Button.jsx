import React from 'react';
import PropTypes from 'prop-types';

import { bemCls } from '../../utils/ClassNameHelpers';

import './Button.css';

const Button = ({
  className, text, type, disabled, onClick,
}) => {
  if ((onClick || type === 'submit') && text) {
    return (
      // eslint-disable-next-line react/button-has-type
      <button
        type={type}
        className={`Button ${className} ${
          bemCls(`Button ${className}`, '--disabled', disabled)}`}
        onClick={onClick}
        disabled={disabled}
      >
        {text
          && (
            <div className={`${bemCls(`Button ${className}`, '__text')} ${
              bemCls(`Button ${className}`, '__text--disabled', disabled)}`}
            >
              {text}
            </div>
          )
        }
      </button>
    );
  }
  return null;
};

export default Button;

Button.propTypes = {
  className: PropTypes.string,
  text: PropTypes.string,
  type: PropTypes.string,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  className: '',
  text: '',
  type: 'button',
  disabled: false,
  onClick: undefined,
};
