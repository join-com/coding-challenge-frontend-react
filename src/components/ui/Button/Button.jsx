import React from "react";
import PropTypes from "prop-types";
import StyledButton from "./StyledButton";

const Button = ({ children, uppercase, onClick, size, type, disabled }) => (
  <StyledButton size={size} uppercase={uppercase} onClick={onClick} type={type} disabled={disabled}>
    {children}
  </StyledButton>
);
Button.propTypes = {
  /** Button content */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  /** Button size */
  size: PropTypes.oneOf(["small", "normal", "large"]),
  /** Is text uppercase */
  uppercase: PropTypes.bool,
  /** on click callback */
  onClick: PropTypes.func,
  /** button type */
  type: PropTypes.string,
  /** is button disabled */
  disabled: PropTypes.bool
};
Button.defaultProps = {
  size: "normal",
  type: "button",
  uppercase: true,
  onClick: () => {}
};
export default Button;
