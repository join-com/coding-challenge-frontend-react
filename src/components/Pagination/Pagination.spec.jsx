import React from "react";
import { shallow, mount } from "enzyme";
import Component from "./Pagination";
import Button from "../ui/Button/Button";

describe("Pagination component test", () => {
  const defaultProps = {
    pages: [1, 2, 3],
    currentPage: 0,
    changeUi: jest.fn()
  };

  const getWrapper = (testProps = {}) => {
    const props = { ...defaultProps, ...testProps };
    return shallow(<Component {...props} />);
  };

  it("should match snapshot ", () => {
    const wrapper = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it("should correct handle pagination buttons click only on not disabled buttons", () => {
    const changeUi = jest.fn();
    const props = { ...defaultProps, changeUi };
    const wrapper = mount(<Component {...props} />);

    const buttons = wrapper.find(Button);
    expect(buttons).toHaveLength(7);
    expect(buttons.at(0).props().disabled).toBeTruthy();
    expect(buttons.at(1).props().disabled).toBeTruthy();
    expect(buttons.at(2).props().disabled).toBeTruthy();
    expect(buttons.at(3).props().disabled).toBeFalsy();
    expect(buttons.at(4).props().disabled).toBeFalsy();
    expect(buttons.at(5).props().disabled).toBeFalsy();
    expect(buttons.at(6).props().disabled).toBeFalsy();

    // shouldn't handle click on disabled button
    buttons.at(0).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(0);

    buttons.at(3).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(1);
    expect(changeUi).toHaveBeenCalledWith({ name: "currentPage", value: 1 });
  });

  it("should disable 3 last buttons if last page selected", () => {
    const changeUi = jest.fn();
    const props = { ...defaultProps, changeUi, currentPage: 2 };
    const wrapper = mount(<Component {...props} />);

    const buttons = wrapper.find(Button);
    expect(buttons).toHaveLength(7);
    expect(buttons.at(0).props().disabled).toBeFalsy();
    expect(buttons.at(1).props().disabled).toBeFalsy();
    expect(buttons.at(2).props().disabled).toBeFalsy();
    expect(buttons.at(3).props().disabled).toBeFalsy();
    expect(buttons.at(4).props().disabled).toBeTruthy();
    expect(buttons.at(5).props().disabled).toBeTruthy();
    expect(buttons.at(6).props().disabled).toBeTruthy();

    // shouldn't handle click on disabled button
    buttons.at(6).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(0);

    buttons.at(1).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(1);
    expect(changeUi).toHaveBeenCalledWith({ name: "currentPage", value: 1 });
  });

  it("should disable only current page button", () => {
    const changeUi = jest.fn();
    const props = { ...defaultProps, changeUi, currentPage: 1 };
    const wrapper = mount(<Component {...props} />);

    const buttons = wrapper.find(Button);
    expect(buttons).toHaveLength(7);
    expect(buttons.at(0).props().disabled).toBeFalsy();
    expect(buttons.at(1).props().disabled).toBeFalsy();
    expect(buttons.at(2).props().disabled).toBeFalsy();
    expect(buttons.at(3).props().disabled).toBeTruthy();
    expect(buttons.at(4).props().disabled).toBeFalsy();
    expect(buttons.at(5).props().disabled).toBeFalsy();
    expect(buttons.at(6).props().disabled).toBeFalsy();

    // shouldn't handle click on disabled button
    buttons.at(3).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(0);

    buttons.at(1).simulate("click");
    expect(changeUi).toHaveBeenCalledTimes(1);
    expect(changeUi).toHaveBeenCalledWith({ name: "currentPage", value: 0 });
  });

  it("should match snapshot for empty pages ", () => {
    const wrapper = getWrapper({ pages: [] });
    expect(wrapper).toMatchSnapshot();
  });
});
