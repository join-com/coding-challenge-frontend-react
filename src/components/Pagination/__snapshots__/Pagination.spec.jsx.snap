// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component test should match snapshot  1`] = `
<Fragment>
  <div>
    <Button
      disabled={true}
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      « First
    </Button>
    <Button
      disabled={true}
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      ‹ Prev
    </Button>
    <Button
      disabled={true}
      key="1"
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      1
    </Button>
    <Button
      disabled={false}
      key="2"
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      2
    </Button>
    <Button
      disabled={false}
      key="3"
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      3
    </Button>
    <Button
      disabled={false}
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      Next ›
    </Button>
    <Button
      disabled={false}
      onClick={[Function]}
      size="normal"
      type="button"
      uppercase={true}
    >
      Last »
    </Button>
  </div>
</Fragment>
`;

exports[`Pagination component test should match snapshot for empty pages  1`] = `<Fragment />`;
